src/web/tracing.rs

```diff
@@ -42,7 +46,71 @@ impl Fairing for RequestIdHeader {
         if let Some(request_id) = request.guard::<&RequestId>().succeeded() {
             response.set_raw_header("X-Request-Id", request_id.0.clone());
         } else {
-            eprintln!("Failed to generate a request id");
+            error!("Failed to generate a request id");
+        }
+    }
+}
+
+/// Rocket fairing that creates and destroys a request `span`.
+///
+/// tracing framework uses a notion of `spans` to store contextual information
+/// alongside log events. For web applications a typical span would be a single
+/// HTTP request, which has a beginning and end time, unique request identifier,
+/// etc.
+///
+/// A natural point of integration with Rocket is a fairing, which allows us to
+/// execute some code before and after each HTTP request, which is exactly what
+/// we need for determining the boundaries of a request span.
+pub struct RequestSpan;
+
+// It is extremely annoying that we need to resort to storing the span guard in
+// a thread local instead of the Rocket Request local cache, but those two have
+// incompatible requirements:
+//
+//  * EnteredSpan is marked as !Send, so that it's not possible to move it
+//    between threads, as that would produce incorrect traces
+//
+//  * Request::local_cache() requires the type to be Send, even though the
+//    values are not actually transferred between threads
+//
+// Using a thread-local is fine, because both on_request()/on_response() and the
+// actual request handler are executed in the same thread.
+thread_local!(static REQUEST_SPAN: RefCell<Option<EnteredSpan>> = RefCell::new(None));
+
+impl Fairing for RequestSpan {
+    fn info(&self) -> Info {
+        Info {
+            name: "Request span",
+            kind: Kind::Request | Kind::Response,
+        }
+    }
+
+    fn on_request(&self, request: &mut Request, _: &Data) {
+        if let Some(RequestId(request_id)) = request.guard::<&RequestId>().succeeded() {
+            let span = info_span!(
+                "request",
+                id = request_id.as_str(),
+                method = request.method().as_str(),
+                uri = request.uri().to_string().as_str(),
```

Comment:
Do I understand correctly that these 3 are going to be repeated all over all dropped logs within a single HTTP request?

Reply:
Correct. It looks like this now:

```
Mar 06 21:33:03.750  INFO request{id="241137c1-79f1-4d2d-a5a0-30bdeaa680ff" method="GET" uri="/v1/snippets/1ssZckPV" status=200}: xsnippet_api::web::tracing: request succeeded
Mar 06 21:33:06.545  INFO request{id="17feead9-56cf-4180-8506-d42f8b7b02ff" method="GET" uri="/v1/snippets/1ssZckPV" status=200}: xsnippet_api::web::tracing: request succeeded
Mar 06 21:33:17.249 DEBUG request{id="a3858f9a-3257-41a1-8624-9b1ad3cf32de" method="GET" uri="/v1/snippets/1ssZckPa"}: xsnippet_api::errors: ApiError: NotFound("Snippet with id `1ssZckPa` is not found")
Mar 06 21:33:17.249  WARN request{id="a3858f9a-3257-41a1-8624-9b1ad3cf32de" method="GET" uri="/v1/snippets/1ssZckPa" status=404}: xsnippet_api::web::tracing: request failed
Mar 06 21:33:30.214  INFO request{id="44abdec6-8046-44fb-a909-57d5f02b789d" method="POST" uri="/v1/snippets" status=201}: xsnippet_api::web::tracing: request succeeded
```